pynamodb/attributes.py

```diff
@@ -860,31 +857,17 @@ def _get_serialize_class(cls, key, value):
         return _get_class_for_serialize(value)
 
     @classmethod
-    def _get_deserialize_class(cls, key, value):
-        if not cls.is_raw():
-            return cls.get_attributes().get(key)
-        return _get_class_for_deserialize(value)
-
-
-def _get_value_for_deserialize(value):
-    key = next(iter(value.keys()))
-    if key == NULL:
-        return None
-    return value[key]
-
-
-def _get_class_for_deserialize(value):
-    value_type = next(iter(value.keys()))
-    if value_type not in DESERIALIZE_CLASS_MAP:
-        raise ValueError('Unknown value: ' + str(value))
-    return DESERIALIZE_CLASS_MAP[value_type]
+    def _get_deserialize_class(cls, attr_name, attr_type):
+        if not cls.is_raw() and attr_type != NULL:
+            return cls.get_attributes().get(attr_name)
+        return DESERIALIZE_CLASS_MAP[attr_type]
 
 
 def _get_class_for_serialize(value):
     if value is None:
         return NullAttribute()
     if isinstance(value, MapAttribute):
-        return type(value)()
+        return value
```

Comment:
Serialize subclasses using the instance stored in the attribute_values dictionary instead of instantiating a new one.

Reply:
(I wonder why it was done this way in the first place)